### PR TITLE
Optimise `import qiskit` with lazy imports

### LINKED ISSUE #7498: DEBUG log messages emitted on imports
### Environment

- **Qiskit Terra version**: main (but previous release likely also have the issues
- **Python version**:  3.10
- **Operating system**:  Linux

### What is happening?

When importing some modules, including the root `qiskit` namespace, `DEBUG` log messages get emitted. For example

`import qiskit` emits:

```
DEBUG:qiskit.quantum_info.synthesis.two_qubit_decompose:Requested fidelity: None calculated fidelity: 1.0 actual fidelity 0.9999999999999997
```

and `import qiskit.test` emits:

```
DEBUG:qiskit.test.testing_options:{'skip_online': False, 'mock_online': False, 'run_slow': False, 'rec': False}
```

### How can we reproduce the issue?

```python3
import logging
logging.basicConfig(level="DEBUG")
import qiskit
```

### What should happen?

We shouldn't be emitting runtime log messages during an import. This is primarily a side effect of code being executed at the module level on import, normally to setup global objects, but emitting these log messages on import is missing the context and really serves no useful purpose.

### Any suggestions?

First we need to find all the places this is happening in the tree (I only tested these two paths there might be others). Then we either remove the logging if it serves no purpose (arguable the qiskit.test example falls in this category )or filter the logs from code that is executed on import (the decomposition falls in this example because normally logging the decomposition fidelity can be useful when code is explicitly calling the synthesis routines).

### PR DESCRIPTION
**Pull request by jakelishman**
### Summary

- Speed up `import qiskit` significantly - 1.17s to 0.57s (2x faster) for a Python environment with almost all the optional dependencies, and 0.78s to 0.52s (33% faster) for a Python environment with _only_ Terra and its requirements.
- Add a new module `qiskit.utils.optionals`, with many lazy-evaluation `HAS_X` variables (e.g. `HAS_MATPLOTLIB`, `HAS_AER`), which can be used for lazy testing of optional dependencies, and unified handling of errors if they are not present.

### Details and comments

These two commits significantly speed up `import qiskit` by deferring many imports to their call sites.  In particular, no component of Scipy is loaded, nor is `pkg_resources` (which is surprisingly heavy).  I didn't change any `scipy` imports in algorithms or opflow, because these aren't imported by default, and I thought they might be a bit more work for lesser benefit.

The first commit in this patch also introduces many new `HAS_X` variables, to unify handling of optional dependencies all the way through the codebase.  Previously, many modules were doing ad-hoc testing with
```python
try:
    import optional_dependency
    HAS_OPTIONAL = True
except ImportError:
    HAS_OPTIONAL = False

class MyClass:
    def __init__(self):
        if not HAS_OPTIONAL:
            raise MissingOptionalLibraryError(...)
```

This form has a couple of downsides:

1. We try to import the optional dependency at import time, which adds to the weight of `import qiskit` if the dependency is installed, even if it won't be used.
2. It clutters many functions with error-handling code, even though they are doing the same thing.
3. Several modules were repeating the exact same code in a couple of places.

The first commit introduces some new lazy-loader objects (`LazyImportTester` and `LazySubprocessTester`), which can be evaluated as Booleans, and will lazily test the presence of the dependencies.  This is similar to the old `HAS_MATPLOTLIB` variable.  There is a new module `qiskit.utils.optionals` that exposes many `HAS_X` variables of these types.  In addition to the lazy Boolean checkers, they also come with unified error handling: modules now can do
```python
@HAS_MATPLOTLIB.require_in_call
def needs_matplotlib(self):
    import matplotlib

def chooses_based_on_argument(self, arg):
    if arg == "text":
        ...
    elif arg == "matplotlib":
        HAS_MATPLOTLIB.require_now("matplotlib drawer")
        import matplotlib
        ...
```

Both `require` forms will raise a suitable `MissingOptionalLibraryError` if the dependency is not available, at the point of call - the decorator raises when the inner function is called, not when the decorator is applied.

Moving the variables into `qiskit.utils.optionals` is also nice for the test suite - instead of it having to repeat the code or import names from random locations, there's a simple namespace where all the objects exist (with documentation on each one in Sphinx).

Fix #7498, as a side-effect - all the instances where initialisation code was logging are now deferred to their usage sites, so the logs will actually make sense and be useful.  You can import every single module defined in Terra without us issuing a log message now.

### CODE REVIEW

**Comment by woodsp-ibm** [comment]
I guess one downside of the new way, from a user perspective, I write code with a bunch of imports and it all seems fine as it loads and starts running. But instead of it failing fast as it used to, I now find it fails some time later if some optional dependency is not present when it finally goes to construct/uses something. The code may have expended significant computational resources by that point. And I might fix that and run again, go further and hit another missing optional. The fail early that was done before would catch the optional dependence early; but has other considerations that you point out such as loading resources that are never used in practice.

**Comment by jakelishman** [comment]
This doesn't actually change anything about that - before, all the imports would silently fail immediately, setting (e.g.) `_HAS_SKQUANT = False`, and then you wouldn't get an error til you instantiated the class.  In this form, you still get the error in the exact same place, but you also don't pay the import cost til then either.

**Comment by woodsp-ibm** [comment]
Ah right, as you say the behavior in terms of it failing later is unaltered by your change. I know in Aqua we used to indicate the failure to load but guess that was removed as it may have been too noisy if you did not care about installing the optional.

**Comment by jakelishman** [comment]
Yeah, I don't know when it changed, but it was probably the right call - this PR identified 26 different optional dependencies across Terra alone.  The code from Aqua seems to be responsible for about 4 of them, which is a bit more reasonable, but still, they're pretty niche optimisers, and I'd imagine most people wouldn't be bothered.  In general, you should get the error when you instantiate the optimiser classes, which is likely going to be one of the first things you do.

File: qiskit/opflow/gradients/gradient.py
Comment (mtreinish) [inline qiskit/opflow/gradients/gradient.py:196]: Is there a reason to not convert this if/else into the `require_now()` usage?

File: qiskit/opflow/gradients/hessian.py
Comment (mtreinish) [inline qiskit/opflow/gradients/hessian.py:241]: Same question about `require_now()` here as in gradient.py or decorating the `get_hessian()` method.

File: qiskit/opflow/gradients/hessian.py
Comment (mtreinish) [inline qiskit/opflow/gradients/hessian.py:258]: Same question about `require_now()` here as in gradient.py

File: qiskit/quantum_info/synthesis/two_qubit_decompose.py
Comment (mtreinish) [inline qiskit/quantum_info/synthesis/two_qubit_decompose.py:1411]: Oh, too bad I was excited when I opened #7498 that we'd have the first use case for module level getattr now that we're >=3.7 but I guess not. Having this comment here is useful for explaining this I should have added one for the lazy aer and ibmq wrappers

File: qiskit/test/base.py
Comment (mtreinish) [inline qiskit/test/base.py:45]: Does this error if you have fixtures installed but not testtools? I don't think this will ever come up because testtools is a requirement of fixtures: https://github.com/testing-cabal/fixtures/blob/master/requirements.txt (it's actually a bidirectional dependency much to the annoyance of linux package maintainers everywhere: https://github.com/testing-cabal/testtools/blob/master/requirements.txt#L5 ). But I'm not sure it would be correct even if it did come up because if we're missing fixtures or testtools we shouldn't error we just shouldn't be using the class with fixtures.

File: qiskit/transpiler/passes/__init__.py
Comment (mtreinish) [inline qiskit/transpiler/passes/__init__.py:82]: This is a good catch but kind of an unrelated change.

File: qiskit/utils/lazy_tester.py
Comment (mtreinish) [inline qiskit/utils/lazy_tester.py:111]: I'm wondering if we should have a class decorator like this too `require_for_instance` that will wrap the class's `__init__` with a require now. Not sure it makes much of a difference, but there is a lot of places where we manually call `require_now()` in the `__init__` but before the super and I'm wondering if having a simpler pattern like
```
@_optionals.HAS_Z3.require_for_instance
class FancyTranspilerPass(TransformationPass):
```

might be less error prone. Although it's been a while since I wrote a class decorator so it might not be trivial.

File: qiskit/utils/optionals.py
Comment (mtreinish) [inline qiskit/utils/optionals.py:292]: Is there a reason we delete these from the module here? It's a bit weird could we just use `__all__` instead? If it's not sufficient could you add a comment explaining why this is needed?

File: qiskit/visualization/__init__.py
Comment (mtreinish) [inline qiskit/visualization/__init__.py:130]: Is there a reason to not deprecate these now?

File: qiskit/visualization/__init__.py
Comment (mtreinish) [inline qiskit/visualization/__init__.py:131]: What about `HAS_PDFTOCAIRO` and `HAS_PIL`?

**Review by mtreinish — COMMENTED** [review]
Awesome thanks for doing this. The API around managing optional dependencies is so much cleaner here and much easier to manage than the ad-hoc solution we were doing before. It's nice enough that I'm slightly tempted to violate the stable branch policy so we can start using this in downstream qiskit packages sooner rather than later :) That coupled with fixing the stray debug logging and the import time improvements this really nice!

Just have a few questions and comments inline

File: qiskit/utils/optionals.py
Comment (jakelishman) [reply qiskit/utils/optionals.py]: It was just me being picky about not leaking names out of the module.  Anything we make public in a module _invariably_ seems to end up with people using that as the import path, so I was just tidying up the names.  I changed that to be more explicit in 897f657db by making the names private right from the import.

File: qiskit/visualization/__init__.py
Comment (mtreinish) [reply qiskit/visualization/__init__.py:130]: Wow, something is still using `qiskit.tools.visualization` we made that move like 3 years ago. :)

**Comment by jlapeyre** [comment]
A couple of lines in `CONTRIBUTING.md` might be useful. Something about when to add a dependency to  `requirements.txt` and when/how to add it to `optionals.py`.
(This looks so useful, I'm surprised there's no package for it, or at least how-to for something similar)

**Comment by jakelishman** [comment]
Added a section into `CONTRIBUTING` in fd74223.

The dependency-checker classes themselves I suppose could be split off into a separate package, but there would be some additional work around making the error handling nice and extensible there.  I think the majority of the benefit comes from defining all the `HAS_X` variables, and those would still need to be written by each project, as in `utils.optionals`.

File: qiskit/utils/optionals.py
Comment (mtreinish) [inline qiskit/utils/optionals.py:101]: Besides the latex drawer, it's also used by the dag visualizations and anywhere we call to pydot or graphviz to wrap the generated image file in an object we can return.

File: test/python/opflow/test_gradients.py
Comment (mtreinish) [inline test/python/opflow/test_gradients.py:479]: A potential follow-on, not something we should do in this pr. Is to add a skip test decorator to the lazy importer class. So we can just make this `@optionals.HAS_JAX.test_requires` or something

File: test/python/utils/test_lazy_loaders.py
Comment (mtreinish) [inline test/python/utils/test_lazy_loaders.py:74]: Is there an advantage to this vs L71? This should be equivalent to `self.assertFalse(test_generator())`

**Review by mtreinish — APPROVED** [review]
This LGTM. I think everything is covered here. Thanks for doing this. I left a few questions/comments inline but nothing work blocking this over. Some things we can always adjust in follow ups as necessary, especially given the size of this PR it'll be good to merge this sooner rather than later.

**Comment by mtreinish** [comment]
Marking this as on hold until we get all the 0.19.2 PRs backported (just to avoid potential merge conflicts). After we're ready to release 0.19.2 I'll tag this as automerge.

File: qiskit/utils/classtools.py
Comment (jakelishman) [reply qiskit/utils/classtools.py:70]: They come up far more than you'd expect - things like wrapping method from `type` and `object` are part of the core usage - wrapping a non-existent `__init__` needs to access something from `object`, and wrapping `__init_subclass__` (which we do in the test classes) needs to access something from `type`.  Wrapping `classmethod` and `staticmethod` instances I think already happens in what we use it for now.

Using builtins as the wrappers happened when I was designing the class in the interpreter - I tried to use `print` as a wrapper for a really poor man's debugger.

File: test/python/opflow/test_gradients.py
Comment (jakelishman) [reply test/python/opflow/test_gradients.py:479]: I'm not 100% sure on that - it would couple the objects to our testing framework, and that's not ideal for something we're publicly exporting.

File: qiskit/transpiler/passes/routing/algorithms/bip_model.py
Comment (jakelishman) [reply qiskit/transpiler/passes/routing/algorithms/bip_model.py:33]: Yeah, this was purely a mechanical change - I didn't do much thinking about where errors should occur, I just tried to keep the pre-existing behaviour wherever it was already lazy.

File: qiskit/utils/optionals.py
Comment (jakelishman) [reply qiskit/utils/optionals.py:101]: Thanks - I rewrote it a little bit to make it clearer that it's used in a few places.

3dbbe56ee

File: test/python/utils/test_lazy_loaders.py
Comment (jakelishman) [reply test/python/utils/test_lazy_loaders.py:74]: Yeah, it's a slightly superfluous test - I'm not 100% sure what I was thinking, but I suspect it was along the lines of "check a few possible implicit Boolean contexts".  Strictly, all three tests here should be identical, I think.  Happy to remove it (and the similar test on line 65 if you'd prefer to reduce noise.

File: qiskit/transpiler/passes/optimization/hoare_opt.py
Comment (jakelishman) [reply qiskit/transpiler/passes/optimization/hoare_opt.py:46]: Yep, that's an accidental change.  I've called it out in the release notes, and updated the docstring that I can now see that I missed (I have docstrings folded in vim by default, so I miss them sometimes).

0cef1576f

File: test/python/opflow/test_gradients.py
Comment (mtreinish) [reply test/python/opflow/test_gradients.py:479]: Well I was thinking more just having the decorator raise `unittest.SkipTest` which is in stdlib and while it's specifically part of unittest framework I think all the python testing frameworks support it. But it's not critical and I agree a gray area of whether we want that as part of the api. We can discuss it later if we think it's something worthwhile

File: qiskit/utils/__init__.py
Comment (mtreinish) [reply qiskit/utils/__init__.py:72]: Nah, this is fine. I was just curious.

**Review by mtreinish — APPROVED** [review]

**Pull request merged**
